Demonstrate Social UI with additional plugins

benjie opened the pull request:
## Description

This PR demonstrates how to utilise additional plugins with a showcase (or other) editor. In this case we've enhanced the "Social UI" editor to support some rich text features, but it demonstrates well how to do this without requiring changes to any `@remirror/*` core module.

Included are some improvements to the documentation to make the API more obvious for people new to the codebase.

## Checklist

- [x] I have read the [**contributing**][contributing] document.
- [x] My code follows the code style of this project and `yarn fix` runs successfully.
- [x] I have updated the documentation where necessary.
- ~~New code is unit tested and all current tests pass when running `yarn test`.~~

## Screenshots

![Screenshot_20200109_122638](https://user-images.githubusercontent.com/129910/72067695-4ed48a80-32db-11ea-98fb-e9692166bda0.png)


[contributing]: https://github.com/ifiokjr/remirror/blob/master/docs/pages/contributing.md

benjie commented on @remirror/editor-social/src/components/social-editor.tsx:
```diff
@@ -394,6 +425,26 @@ export class SocialEditor extends PureComponent<SocialEditorProps, State> {
+            />
+          ) : null}
+          {rich ? <RemirrorExtension Constructor={SSRHelperExtension} /> : null}
+
```
I'm pretty sure this is achieveable by passing the `extensions` prop to `<SocialEditor` directly and having all this extension code instead live in `examples/with-next/pages/editor/rich-social/index.tsx`. That way we can demonstrate how to add extensions to a showcase editor without requiring changes to core.

However, I cannot figure out what the equivalent of these extension components is when passing to the `extensions` prop. Please can you advise?

Also, would it make sense to allow the `extensions` prop to accept these `<RemirrorExtension />` elements for consistency with the React API?

ifiokjr commented:
Looks good so far @benjie 
I'll take a look later today.

benjie commented on @remirror/showcase/tsconfig.prod.json:
```diff
@@ -5,7 +5,8 @@
     "composite": true,
-    "emitDeclarationOnly": true,
+    "emitDeclarationOnly": false,
+    "module": "commonjs",
```
:rotating_light: DON'T MERGE THIS

I don't know why this seems required; I think I need catching up on the build system (and once caught up I can write up some instructions for others).

ifiokjr replied:
Sure thing, I'll hold.

benjie commented on docs/pages/contributing.md:
```diff
@@ -176,3 +176,7 @@ available at [http://contributor-covenant.org/version/1/4][version]
+
+### Troubleshooting
+
+If you're getting errors like `ReferenceError: CodeBlockExtension is not defined` but you know you've imported it, it might be because you've not added it as a dependency to the relevant `package.json`. Rather than throwing an error in this case, rollup (?) seems to just drop the import statement but still persist the lines where the import is used.
```
This had me scratching my head for a good while! Is it rollup doing this, or...?

ifiokjr replied:
I ran into the problem a while back. I think I fixed it changing `export { Value } from './file` to `export * from './file'`. I don't think the babel build can currently determine which items to export from a typescript file since it's not sure which is a type and which is an actual value (when directly exporting).

I could be completely wrong though.

benjie replied:
Interesting. For me, adding the dependencies to package.json solved it.

ifiokjr replied:
Brilliant! 😄 

I might be conflating two unrelated issues to be honest.

benjie commented:
(Rebased onto #187)

benjie commented:
(Rebased onto canary.)

When you get a little time @ifiokjr, I'd appreciate some help understanding:

1. the `extensions` prop to `RemirrorManager` (and how to convert between values for that and values for child `<RemirrorExtension />` elements)
2. the build system, and why I seem to have had to the change to the `tsconfig` above - am I running the wrong command (`yarn dev`)?

ifiokjr commented:
> 2. The build system, and why I seem to have had to the change to the tsconfig above - am I running the wrong command (yarn dev)?

I'll checkout the repo and get back to you.

> 3. CI, I've never used Azure pipelines before and I'm finding it hard to figure out what I did wrong

This completely my fault. I'm using the newest version of yarn which produces a different `yarn.lock` file causing the `yarn integrity` check to fail on the CI server. I've removed now that check since it's not actually important.

ifiokjr commented:
Instead of using yarn dev I prefer to use storybooks while developing then navigate to http://localhost:6006

```bash
yarn storybook
```

To add a new story create a file with this structure - `__stories__/<FILE_NAME>.stories.tsx`. It will automatically be picked up. [This folder](https://github.com/ifiokjr/remirror/tree/b096ed1/%40remirror/showcase/src/__stories__) is a good example.

The reason that yarn dev wasn't working for you is that the project is using composite types and adding a new dependency to the project throws the build process since it location has to explicitly be updated. Luckily there's a command for that. 

Running the following will automatically update all your project references so that the build still works. 

```bash
yarn generate:json
```

It basically creates all the project `tsconfig.prod.json` files for you as can be seen [here](https://github.com/ifiokjr/remirror/blob/b096ed1dd3/support/scripts/generate-configs.js#L186-L228).

ifiokjr commented:
Also if you rebase now, the build should work properly for you.

benjie commented:
It'd be helpful if the size limit checks called out the failing package more boldly; having to scan the entire list carefully is a bit of a chore.

```
@remirror/editor-social
  Package size limit has exceeded by 2.44 KB
  Size limit:   200 KB
  Size:         202.44 KB with all dependencies, minified and gzipped
  Loading time: 4.1 s     on slow 3G
```

This is because wysiwyg-style features have been added to editor-social. These might be removed again if I can get to the bottom of (1).

benjie commented:
I'm completely lost in this build system, I don't understand the errors it's giving me.

For example, I changed `const userData ...` to `export const userData ...` in `@remirror/showcase/src/social.tsx` so I can use it from `examples/with-next`.

I then tried `import { userData } from '@remirror/showcase';` from a file in `examples/with-next`.

I then run `yarn dev` in `examples/with-next` and when I load the web page I get:

<details>
<summary>[ error ] ./pages/editor/rich-social/rich.tsx: Attempted import error: 'userData' is not exported from '@remirror/showcase'.
Could not find files for /editor/rich-social/content in .next/build-manifest.json
ModuleDependencyError: "export 'userData' was not found in '@remirror/showcase'</summary>


```
examples/with-next$ yarn dev
yarn run v1.21.0
$ next
[ wait ]  starting the development server ...
[ info ]  waiting on http://localhost:3000 ...
> Using external babel configuration
> Location: "/home/benjie/Dev/Sprout/remirror/examples/with-next/.babelrc.js"
[ info ]  bundled successfully, waiting for typecheck results...
[ wait ]  compiling ...
[ info ]  bundled successfully, waiting for typecheck results...
[ ready ] compiled successfully - ready on http://localhost:3000
[ event ] build page: /editor/rich-social/content
[ wait ]  compiling ...
[ error ] ./pages/editor/rich-social/rich.tsx
Attempted import error: 'userData' is not exported from '@remirror/showcase'.
Could not find files for /editor/rich-social/content in .next/build-manifest.json
ModuleDependencyError: "export 'userData' was not found in '@remirror/showcase'
    at Compilation.reportDependencyErrorsAndWarnings (/home/benjie/Dev/Sprout/remirror/node_modules/next/node_modules/webpack/lib/Compilation.js:1468:21)
    at /home/benjie/Dev/Sprout/remirror/node_modules/next/node_modules/webpack/lib/Compilation.js:1258:10
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/home/benjie/Dev/Sprout/remirror/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:15:1)
    at AsyncSeriesHook.lazyCompileHook (/home/benjie/Dev/Sprout/remirror/node_modules/tapable/lib/Hook.js:154:20)
    at Compilation.finish (/home/benjie/Dev/Sprout/remirror/node_modules/next/node_modules/webpack/lib/Compilation.js:1253:28)
    at /home/benjie/Dev/Sprout/remirror/node_modules/next/node_modules/webpack/lib/Compiler.js:672:17
    at _done (eval at create (/home/benjie/Dev/Sprout/remirror/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:9:1)
    at eval (eval at create (/home/benjie/Dev/Sprout/remirror/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:31:22)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
ModuleDependencyError: "export 'userData' was not found in '@remirror/showcase'
    at Compilation.reportDependencyErrorsAndWarnings (/home/benjie/Dev/Sprout/remirror/node_modules/next/node_modules/webpack/lib/Compilation.js:1468:21)
    at /home/benjie/Dev/Sprout/remirror/node_modules/next/node_modules/webpack/lib/Compilation.js:1258:10
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/home/benjie/Dev/Sprout/remirror/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:15:1)
    at AsyncSeriesHook.lazyCompileHook (/home/benjie/Dev/Sprout/remirror/node_modules/tapable/lib/Hook.js:154:20)
    at Compilation.finish (/home/benjie/Dev/Sprout/remirror/node_modules/next/node_modules/webpack/lib/Compilation.js:1253:28)
    at /home/benjie/Dev/Sprout/remirror/node_modules/next/node_modules/webpack/lib/Compiler.js:672:17
    at _done (eval at create (/home/benjie/Dev/Sprout/remirror/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:9:1)
    at eval (eval at create (/home/benjie/Dev/Sprout/remirror/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:31:22)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
[ info ]  bundled successfully, waiting for typecheck results...
[ ready ] compiled successfully - ready on http://localhost:3000
```
</details>

But

1. I added the export to `social.tsx`
2. `index.tsx` has `export * from './social'` so it should be included

The issue seems to be that the compiled version `@remirror/showcase/lib/index.js` does not contain `userData` (but does contain other changes I've made).

I've tried running `yarn generate:json` and I've tried deleting the `.next` folder but I just can't figure out why the export I have in my code is not making through to the output code.


We've also got another issue which may or may not be related:

```
 → @remirror/editor-social/lib/dist/editor-social.esm.js, @remirror/editor-social/lib/dist/editor-social.cjs.js...
[!] Error: 'default' is not exported by node_modules/refractor/core.js
https://rollupjs.org/guide/en/#error-name-is-not-exported-by-module
@remirror/extension-code-block/lib/dist/extension-code-block.esm.js (13:7)
11: import { setBlockType } from 'prosemirror-commands';
12: import { TextSelection, Plugin } from 'prosemirror-state';
13: import refractor from 'refractor/core';
           ^
14: import { jsx } from '@emotion/core';
15: import SyntaxHighlighter from 'react-syntax-highlighter/dist/cjs/prism-light';
Error: 'default' is not exported by node_modules/refractor/core.js
    at error (/home/benjie/Dev/Sprout/remirror/node_modules/rollup/dist/rollup.js:5340:30)
    at Module.error (/home/benjie/Dev/Sprout/remirror/node_modules/rollup/dist/rollup.js:9730:9)
    at handleMissingExport (/home/benjie/Dev/Sprout/remirror/node_modules/rollup/dist/rollup.js:9644:21)
    at Module.traceVariable (/home/benjie/Dev/Sprout/remirror/node_modules/rollup/dist/rollup.js:10069:17)
    at ModuleScope.findVariable (/home/benjie/Dev/Sprout/remirror/node_modules/rollup/dist/rollup.js:8689:39)
    at FunctionScope.findVariable (/home/benjie/Dev/Sprout/remirror/node_modules/rollup/dist/rollup.js:3025:38)
    at ChildScope.findVariable (/home/benjie/Dev/Sprout/remirror/node_modules/rollup/dist/rollup.js:3025:38)
    at MemberExpression.bind (/home/benjie/Dev/Sprout/remirror/node_modules/rollup/dist/rollup.js:6453:49)
    at CallExpression$1.bind (/home/benjie/Dev/Sprout/remirror/node_modules/rollup/dist/rollup.js:3112:23)
    at CallExpression$1.bind (/home/benjie/Dev/Sprout/remirror/node_modules/rollup/dist/rollup.js:6641:15)
```

I'm pretty lost right now; the changes I'm trying to make are very simple!

benjie commented:
Ah; I've fixed it. It was the dependencies issue again. I removed some dependencies during refactoring and everything stopped working but with really confusing/misleading error messages. I think `rollup` must refer to `package.json` rather than using node-style resolution like webpack does. Just straight up dropping parts of the code because certain imports aren't present seems like a really terrible failure mode - better to throw an error and abort!

benjie commented:
@ifiokjr This PR now much more clearly demonstrates what I'm attempting to achieve. The issue I have is that I don't understand how to take my `extensionElements` and turn them into something that I can feed into the `extensions` prop. Once I can do that, we can demonstrate how to customise these showcase editors without needing changes to core which is pretty compelling!

Can you help on this? (It's the same as question 1 from last week.)

benjie commented:
> examples/with-next/pages/editor/rich-social/rich.tsx:85:9 - error TS17017: JSX fragment is not supported when using an inline JSX factory pragma

Well... it works. But apparently it doesn't pass TypeChecking.

benjie commented:
![Screenshot_20200116_163500](https://user-images.githubusercontent.com/129910/72543793-279f2f80-387e-11ea-9e91-758122f371e5.png)

benjie commented on @remirror/showcase/src/social.tsx:
```diff
@@ -37,6 +37,8 @@ const userData: UserData[] = fakeUsers.results.map(
   }),
 );
 
+export { fakeUsers, fakeTags, userData };
```
This is the only change that should be necessary to the `@remirror` folder; I'd like to find a way to remove the need for the above changes.

benjie commented on @remirror/react/src/components/remirror-manager.tsx:
````diff
@@ -27,6 +27,32 @@ export interface RemirrorManagerProps {
+   *     priority: priority || 2,
+   *   },
+   * ];
+   * ```
````
![Screenshot_20200116_171211](https://user-images.githubusercontent.com/129910/72546845-67b4e100-3883-11ea-978b-c4f8d394f00a.png)

^ This I think will be very helpful for people :smiley:

benjie commented on packages/multishift/src/multishift-utils.ts:
```diff
@@ -262,7 +262,7 @@ export const callChangeHandlers = <GItem = any>(
  */
 export const getElementIds = (
-  defaultId: string,
+  defaultId: string | number,
```
This change also in #199

benjie commented:
Issue description updated; ready for review. This should definitely be squashed before being merged - I've not done that yet in case you're interested to follow where I got stuck.
